controllers/controller/devworkspacerouting/suite_test.go

```diff
@@ -0,0 +1,191 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devworkspacerouting_test
+
+import (
+	"context"
+	"fmt"
+	"go/build"
+	"os"
+	"path/filepath"
+	"testing"
+
+	dwv1 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha1"
+	dwv2 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	controllerv1alpha1 "github.com/devfile/devworkspace-operator/apis/controller/v1alpha1"
+	"github.com/devfile/devworkspace-operator/controllers/controller/devworkspacerouting"
+	"github.com/devfile/devworkspace-operator/controllers/controller/devworkspacerouting/solvers"
+	routev1 "github.com/openshift/api/route/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/yaml"
+
+	"github.com/devfile/devworkspace-operator/pkg/cache"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/devfile/devworkspace-operator/pkg/infrastructure"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+var (
+	k8sClient         client.Client
+	testEnv           *envtest.Environment
+	ctx               context.Context
+	cancel            context.CancelFunc
+	testControllerCfg = config.GetConfigForTesting(&controllerv1alpha1.OperatorConfiguration{
+		Workspace: &controllerv1alpha1.WorkspaceConfig{},
+		Routing: &controllerv1alpha1.RoutingConfig{
+			ClusterHostSuffix: "test-environment-cluster-suffix",
+		},
+		EnableExperimentalFeatures: pointer.Bool(true),
+	})
+)
+
+func TestAPIs(t *testing.T) {
+	if os.Getenv("SKIP_CONTROLLER_TESTS") == "true" {
+		t.Skip()
+	}
+
+	RegisterFailHandler(Fail)
+
+	RunSpecs(t, "DevWorkspaceRouting Controller Suite")
+}
+
+var _ = BeforeSuite(func() {
+	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+
+	ctx, cancel = context.WithCancel(context.TODO())
+
+	By("setting up controller environment")
+	Expect(setupEnvVars()).To(Succeed())
+
+	By("bootstrapping test environment")
+	testEnv = &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join("..", "..", "..", "deploy", "templates", "crd", "bases"),
+			filepath.Join(build.Default.GOPATH, "pkg", "mod", "github.com", "openshift", "api@v0.0.0-20220729142910-83d1191dd9fc", "route", "v1", "route.crd.yaml"),
```

Comment:
There's currently an error on the CI related to this [workaround](https://github.com/kubernetes-sigs/controller-runtime/issues/1191#issuecomment-769678481):

```go
  Unexpected error:
      <*fmt.wrapError | 0xc0005f84c0>: {
          msg: "unable to install CRDs onto control plane: unable to read CRD files: stat /home/runner/go/pkg/mod/github.com/openshift/api@v0.0.0-20220729142910-83d1191dd9fc/route/v1/route.crd.yaml: no such file or directory",
          err: <*fmt.wrapError | 0xc0005f84a0>{
              msg: "unable to read CRD files: stat /home/runner/go/pkg/mod/github.com/openshift/api@v0.0.0-20220729142910-83d1191dd9fc/route/v1/route.crd.yaml: no such file or directory",
              err: <*fs.PathError | 0xc00051fcb0>{
                  Op: "stat",
                  Path: "/home/runner/go/pkg/mod/github.com/openshift/api@v0.0.0-20220729142910-83d1191dd9fc/route/v1/route.crd.yaml",
                  Err: <syscall.Errno>0x2,
              },
          },
      }
      unable to install CRDs onto control plane: unable to read CRD files: stat /home/runner/go/pkg/mod/github.com/openshift/api@v0.0.0-20220729142910-83d1191dd9fc/route/v1/route.crd.yaml: no such file or directory
  occurred
  In [BeforeSuite] at: /home/runner/work/devworkspace-operator/devworkspace-operator/controllers/controller/devworkspacerouting/suite_test.go:93
```

Will have to resolve it.

Reply:
Turns out this is a bit more tricky than I had hoped.
`openshift/api v0.0.0-20200205133042-34f0ec8dab87` doesn't have the route CRD file, and upgrading to `@v0.0.0-20220729142910-83d1191dd9fc` breaks some transitive dependencies. Currently figuring this out.

Reply:
For the moment, I copied the route CRD to the testdata directory, though I'd like to try some more on upgrading the required dependencies to prevent the need for this workaround. If it ends up being too messy, I'll just keep the copy-pasted route CRD.